V1 blog post (#513)
mdx-js/mdx: docs/blog/v1.md

[MDX Blog](/blog)

# MDX goes stable

It’s been a year and a half since the first MDX commit and a year since MDX was first announced at
ZEIT Day.  MDX is a radical paradigm shift in how to write immersive content using components.  It’s
an open, [authorable format](https://johno.com/authorable-format) that makes it _fun_ to write again.

> Over the last year we’ve been working on numerous bug fixes, new features, better parsing, and integration

Reviewer: ```suggestion
Since announcing MDX we’ve been working on numerous bug fixes, new features, better parsing, and integration
```

(reference back to earlier paragraph)

[MDX Blog](/blog)

# MDX goes stable

It’s been a year and a half since the first MDX commit and a year since MDX was first announced at
ZEIT Day.  MDX is a radical paradigm shift in how to write immersive content using components.  It’s
an open, [authorable format](https://johno.com/authorable-format) that makes it _fun_ to write again.

Over the last year we’ve been working on numerous bug fixes, new features, better parsing, and integration
tests.  Now, we think it’s ready.  **We’re happy to finally release v1!**

## 🎉 What’s new?

Here’s a rough breakdown on what we’ve been working on:

### Parsing

We’ve put a lot of work into improving the parsing of MDX documents.  This not only includes better
handling of import and export blocks, but being more flexible with JSX blocks.  We won’t get into the
nitty gritty here, but rest assured we’ve added many more tests and are handling all the edge cases

> we’ve currently encountered (please open an issue if you find a new parse issue 😸).

Reviewer: Attempt at a rewrite, not necessarily better!

```txt
Parsing of MDX documents is significantly improved.
We won’t get into the nitty gritty here, but we’ve seen how MDX is used in real-world projects, integrated the edge cases we came across into our test suite, and now handle JSX, imports, and exports much more intuitively.
Please open an issue if you find a case we haven’t covered! 😸
```

### remark-mdx

`remark-mdx` is the syntactic extension for MDX in remark.  It provides the parsing functionality for
MDX as a _remark plugin_.  It’s a bit meta.  With `remark-mdx` you can parse MDX documents
and perform transformations before writing them back to MDX strings.  This allows tools like prettier

> to use the exact same parser as MDX core by installing the library.

Reviewer: Also an attempt at a rewrite, not necessarily better!

```text
`remark-mdx` is the syntactic extension for MDX in remark.  It provides the parsing functionality for
MDX as a _[remark](#) plugin_.  That sounds a bit meta.  What it means is that before we had the syntax parsing bits _in_ the library (unusable from the outside), and now it’s externalised (usable from the outside).  This is useful if you want to inspect or transform MDX documents.  For example, it allows tools like prettier to use the exact same parser as us by using the new plugin to format MDX documents.  Or you use `remark-mdx` in combination with [remark-lint](#) to check your MDX.
```

### Custom pragma

> With v1 we’ve moved away from using `MDXTag` and are using a custom pragma which wraps `React.createElement`.

Reviewer: Should we have an example here? before, after? showing that it’s more legible?
Another random musing: Is there a way we could not name drop React, but keep it vague, so naming Vue later has more effect?

We decided to update this approach so the JSX output is more legible, lighterweight, and more customizable.

> **This means that we can add support for any library that speaks JSX!**

Reviewer: Is `library that speaks JSX` the correct jargon? I’m not sure what the term library expands to here, React, Vue, etc? The term Library is also previously used as `MDX library` relating to our core v. the format, so I’d try and use a different term here. Should it be framework? Pragma?

> Special thanks to [@christopherbiscardi](https://christopherbiscardi.com) for implementing this feature.

Reviewer: Maybe link to GH / Twitter? I’d expect @mentions to go to a social profile.

[Read the blog post](/blog/custom-pragma)

### 📚 Documentation

**Good libraries need great docs**, so we’ve been working on adding content and improving the overall documentation
website experience.

*   [Search (thanks Algolia)](https://mobile.twitter.com/4lpine/status/1114270174096412672)
*   [Guides](https://mdxjs.com)

> *   [Automatic deployment via ZEIT](https://zeit.co)

Reviewer: Should this link be more specific?

> *   [Custom Gatsby theme](https://gatsbyjs.org)

Reviewer: Should this link be more specific?

> *   Reorganization and misc updates

Reviewer: `misc updates` sounds small. To me it feels like the docs are completely rewritten, and I’d suggest someone who previously read the docs to re-read them again, so I would drop this line and mention rewriting in the docs intro paragraph.

Special thanks to [@jxnblk](https://jxnblk.com) and [@wooorm](https://wooorm.com) for their help improving the
docs and updating the build tooling.

## Breaking changes

Unfortunately, we’ve had to introduce a few breaking changes, but we’ve written a migration guide which

> will walk you through the steps needed to upgrade.

Reviewer: This may be very west coast of you, compared to my dutch rudeness, but I wouldn’t start with `Unfortunately,`.
MDX was LTE 1, so I feel like people saw breaking changes coming.
I’d start with something about being significantly better, real-world testing resulting in much better ideas compared to a year ago, and then *even having a migration guide*!

*   🚨`@mdx-js/tag` is replaced by `@mdx-js/react` and an `@mdx` pragma - [migration guide](/migrating/v1#pragma)
*   MDXProvider now merges component contexts when nested - [migration guide](/migrating/v1#mdxprovider)
*   React support now requires `>= 16.8` in `@mdx-js/react` - [migration guide](/migrating/v1#react)

[Read the full v1 migration guide](/migrating/v1)

## Deprecations

> Luckily, we only need to introduce one deprecation.  The `mdPlugins` and `hastPlugins` options have been renamed

Reviewer: This section doesn’t need so much focus, I think it could be a bullet point in the previous section!

`remarkPlugins` and `rehypePlugins` respectively.  For the time being we will issue a warning when the old options
are used.  In v2, the old options will be removed.

## 📈 Growth

A major release is always a good time to take a step back and reflect on what’s been happening in terms of growth
and the greater community.  The ecosystem surrounding MDX has already grown larger than we ever dreamed.  We’ve also
seen projects/products/application we never even imagined.

### Ecosystem

*   [Prettier](https://prettier.io)
*   [Docz](https://docz.site)
*   [MDX Deck](https://github.com/jxnblk/mdx-deck)
*   [Next.js](https://nextjs.org)
*   [Gatsby](https://gatsbyjs.org)
*   [AST Explorer](https://astexplorer.net)
*   [Vue support (alpha)](/vue)
*   [Demoboard](https://frontarm.com/demoboard/)

> *   [Editors](/editors)

Reviewer: The above list is a bit of a mix, they’re cool things, but what makes them be in this list? Vue support is very different from prettier!

Author: Yeah, I kinda arbitrarily added some seemingly relevant things that are new or built around the format 🤷‍♂️

> ### Numbers

Reviewer: This section should be before ecosystem, start with the bang!

*   **Contributors**: 50
*   **Pull requests**: 281
*   **Commits**: 670

The contributor growth is one of the most important aspects here as we have numerous community members that are familiar
with MDX internals.  This allows us to continually improve the project and spread the workload against an ever growing
team of contributors.

[See the contributor graph](https://github.com/mdx-js/mdx/graphs/contributors)

## 🛣 Roadmap

With v1 released we have a roadmap in place that we’ll continue working towards over the next 6 months or
so in addition to bug fixes and parsing issues we might encounter.

*   MDXs: [#454](https://github.com/mdx-js/mdx/issues/454)
*   Interleaving: [#195](https://github.com/mdx-js/mdx/issues/195)
*   Global shortcodes: [#508](https://github.com/mdx-js/mdx/pull/508)
*   Stable Vue support: [#238](https://github.com/mdx-js/mdx/issues/238)
*   Blocks: MDX WYSIWYG: [blocks/blocks](https://github.com/blocks/blocks)
*   MDX playground inspired by AST Explorer: [#220](https://github.com/mdx-js/mdx/issues/220)
*   New splash page: [#444](https://github.com/mdx-js/mdx/issues/444)
*   Showcase page: [#414](https://github.com/mdx-js/mdx/issues/414)

> ### Vue support

Reviewer: This is much cooler than being down here! It’s usable, right? So add it somewhere at the top! And if there are caveats, describe them here and link to it from the top.

Author: My thought on keeping it lower is that it's only an alpha and will have it's own release announcement when more thoroughly tested.

Supporting Vue is an important goal for MDX and is one of the primary reasons we’ve rearchitected MDX to use
custom pragma.  We’re delighted that **we now have an alpha version for Vue working**.  If anyone from the Vue

> community wants to give it a try and provide feedback we’d be eternally grateful.

Reviewer: ```
If anyone from the Vue community wants to give it a try and provide feedback we’d be eternally grateful.
```

Too soft, can be bolder, ask for help! `We’d love it if the Vue community...`
